blockprod/src/detail/mod.rs

```diff
@@ -66,7 +69,141 @@ pub struct JobKey {
 }
 
 pub struct JobHandle {
-    cancel_signal: oneshot::Sender<()>,
+    cancel_sender: UnboundedSender<()>,
+}
+
+pub struct JobManager {
+    chainstate_handle: ChainstateHandle,
+    new_job_sender: UnboundedSender<(Id<GenBlock>, UnboundedSender<()>)>,
+    stop_job_sender: UnboundedSender<(Option<JobKey>, oneshot::Sender<bool>)>,
+    all_jobs: BTreeMap<JobKey, JobHandle>,
+}
+
+impl JobManager {
+    pub fn new(chainstate_handle: ChainstateHandle) -> JobManager {
+        let (new_job_sender, mut new_job_receiver) =
+            mpsc::unbounded_channel::<(Id<GenBlock>, UnboundedSender<()>)>();
+
+        let (stop_job_sender, mut stop_job_receiver) = mpsc::unbounded_channel();
+
+        let mut job_manager = JobManager {
+            chainstate_handle,
+            new_job_sender,
+            stop_job_sender,
+            all_jobs: BTreeMap::new(),
+        };
+
+        let (chainstate_sender, mut chainstate_receiver) =
+            mpsc::unbounded_channel::<Id<GenBlock>>();
+
+        job_manager.subscribe_to_chainstate(chainstate_sender);
+        job_manager.run(chainstate_receiver, new_job_receiver, stop_job_receiver);
+
+        job_manager
+    }
+
+    fn run(
+        &mut self,
+        mut chainstate_receiver: UnboundedReceiver<Id<GenBlock>>,
+        mut new_job_receiver: UnboundedReceiver<(Id<GenBlock>, UnboundedSender<()>)>,
+        mut stop_job_receiver: UnboundedReceiver<(Option<JobKey>, oneshot::Sender<bool>)>,
+    ) {
+        tokio::spawn(async move {
+            let mut all_jobs: BTreeMap<JobKey, JobHandle> = BTreeMap::new();
```

Comment:
While this works for the time being, it'll probably cause an issue later when you try to stop everything on exit. But it's a good start.